### PR TITLE
Fixing the transfer of child shares for transferShare

### PR DESCRIPTION
**Pull request by sharidas**
The child shares are transferred first before deleting
the parent share. If this is not done, then the child
shares are lost, which could result in loss of shares.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

## Description

If the parent share is already shared to the destination user and if the parent share has child shares, then exception would occur while trying to update the child share. Because while processing the parent share, it would have been deleted. And subsequently the child shares would also be deleted.

## Related Issue

https://github.com/owncloud/core/issues/31150

## Motivation and Context

If the parent share is already shared to the destination user and if the parent share has child shares, then exception would occur while trying to update the child share. Because while processing the parent share, it would have been deleted. And subsequently the child shares would also be deleted.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by sharidas** [comment]
Things I have done in this version of PR:
- Process the child shares before the parent shares are getting deleted.
- The usage of query builder inside the `transferShare()` to make the child share decouple from the parent share. That is set it to null.

There is a problem with this solution, if a child share is processed, I need to add a mechanism to stop calling `transferShare` again.

**Comment by sharidas** [comment]
The problem mentioned in https://github.com/owncloud/core/pull/31924#issuecomment-400660863, is rectified in this version.

File: apps/files/lib/Command/TransferOwnership.php [hunk 1]
```diff
@@ -81,12 +85,13 @@ class TransferOwnership extends Command {
...
 	private $finalTarget;
 
-	public function __construct(IUserManager $userManager, IManager $shareManager, IMountManager $mountManager, Manager $encryptionManager, ILogger $logger) {
+	public function __construct(IUserManager $userManager, IManager $shareManager, IMountManager $mountManager, Manager $encryptionManager, ILogger $logger, ProviderFactory $factory) {
 		$this->userManager = $userManager;
 		$this->shareManager = $shareManager;
 		$this->mountManager = $mountManager;
 		$this->encryptionManager = $encryptionManager;
 		$this->logger = $logger;
+		$this->factory = $factory;
```
Comment (PVince81) [inline apps/files/lib/Command/TransferOwnership.php]: factory is a too generic name. please call it `shareProviderFactory` or something similar

File: apps/files/lib/Command/TransferOwnership.php [hunk 2]
```diff
@@ -295,18 +301,39 @@ protected function transfer(OutputInterface $output) {
...
 		$output->writeln("Restoring shares ...");
 		$progress = new ProgressBar($output, \count($this->shares));
+		$status = 0;
 
+		$childShares = [];
 		foreach ($this->shares as $share) {
 			try {
+				/**
+				 * Do not process children which are already processed.
+				 * This piece of code tries to populate the childShare
```
Comment (PVince81) [inline apps/files/lib/Command/TransferOwnership.php]: it doesn't try, it just populates ? "try" would only apply if there are possible expected errors or exceptions

File: apps/files/lib/Command/TransferOwnership.php [hunk 3]
```diff
@@ -295,18 +301,39 @@ protected function transfer(OutputInterface $output) {
...
+				 * And hence will avoid further processing of same share
+				 * again.
+				 */
+				if ($share->getSharedWith() === $this->destinationUser) {
+					$provider = $this->factory->getProviderForType($share->getShareType());
+					foreach ($provider->getChildren($share) as $child) {
+						$childShares[] = $child->getId();
+					}
+				} else {
+					if (\in_array($share->getId(), $childShares, true)) {
```
Comment (PVince81) [inline apps/files/lib/Command/TransferOwnership.php]: not fully clear what this code is for, please add a comment

File: apps/files/lib/Command/TransferOwnership.php [hunk 4]
```diff
@@ -295,18 +301,39 @@ protected function transfer(OutputInterface $output) {
 	private function restoreShares(OutputInterface $output) {
 		$output->writeln("Restoring shares ...");
 		$progress = new ProgressBar($output, \count($this->shares));
+		$status = 0;
 
+		$childShares = [];
```
Comment (PVince81) [inline apps/files/lib/Command/TransferOwnership.php]: I guess this PR is not finished yet ? You don't seem to do anything with the gathered child shares ?

File: lib/private/Share20/Manager.php [hunk 5]
```diff
@@ -717,6 +722,14 @@ public function transferShare(IShare $share, $oldOwner, $newOwner, $finalTarget)
 			throw new NotFoundException("The target location $finalTarget doesn't exist");
 		}
 
+		if ($isChild === true) {
+			//Set the parent to null so that we don't lose the shares after transfer
```
Comment (PVince81) [inline lib/private/Share20/Manager.php]: hmmm, this sounds like the second approach we talked about ? so why gather child shares in the above file then ?
detaching the shares should be enough to avoid them being deleted

**Review by PVince81 — COMMENTED** [review]
Some confusion, possibly you used one approach then changed your mind and did not clean up the old code ?

When setting "parent" to null, how do you know what parent to reattach the share to ? I guess you don't need to reattach as the "file_source" already points at the transferred file id ?

File: lib/private/Share20/Manager.php (hunk 5 above)
Comment (sharidas) [reply lib/private/Share20/Manager.php]: The thought which made me to use `$childShares = []`  is :
- Lets say without this code change in transfer ownership (i.e, without child shares ), and under scenario https://github.com/owncloud/core/issues/31150#issue-314806286, we could see that one of the shares will be called twice. The public link created which is the child of the folder transferred to user2, will be processed https://github.com/owncloud/core/pull/31924/files#diff-a9d8fe189a37b652ea5b821e4d772438R746. Here the public link will be decoupled from the parent. Then again the public link share would be passed in the TransferOwnership.php to the `transferShare()`. So my intention was to remove the duplicate calls to the same share. 

Basically `childShares` will be populated with the children of each share in `$this->share`. My idea is to have share ids to the array. And then while processing I will if the share id is there in the `childShares`. If yes then it wont be transferred. Because it is already transferred.

File: apps/files/lib/Command/TransferOwnership.php (hunk 3 above)
Comment (PVince81) [reply apps/files/lib/Command/TransferOwnership.php]: ah ok, so this is where you use `$childShares`
add a PHP comment to clarify

Comment (sharidas) [reply apps/files/lib/Command/TransferOwnership.php]: Added comment to justify the usage of `$childShares`.

File: apps/files/lib/Command/TransferOwnership.php (hunk 2 above)
Comment (sharidas) [reply apps/files/lib/Command/TransferOwnership.php]: Removed the usage of "try".

File: apps/files/lib/Command/TransferOwnership.php (hunk 1 above)
Comment (sharidas) [reply apps/files/lib/Command/TransferOwnership.php]: Renamed `factory` to `shareProviderFactory` as suggested.

File: apps/files/lib/Command/TransferOwnership.php (hunk 4 above)
Comment (sharidas) [reply apps/files/lib/Command/TransferOwnership.php]: Tested with the latest change and made sure that scenario https://github.com/owncloud/core/issues/31150#issue-314806286 works.

**Comment by sharidas** [comment]
The phpunit is failing in https://github.com/owncloud/core/blob/master/apps/files_sharing/lib/Capabilities.php. This failure doesn't look related to this PR.

**Comment by phil-davis** [comment]
``master`` was broken for a short while - rebase

**Comment by sharidas** [comment]
The reason for 'codecov/patch' not passing could be because of absence of php unit test. All class variables are declared as `private` which is causing `invokePrivate` not to set the values to behave accordingly.

File: apps/files/lib/Command/TransferOwnership.php [hunk 6]
```diff
@@ -295,18 +301,44 @@ protected function transfer(OutputInterface $output) {
...
+				if ($share->getSharedWith() === $this->destinationUser) {
+					$provider = $this->shareProviderFactory->getProviderForType($share->getShareType());
+					foreach ($provider->getChildren($share) as $child) {
+						$childShares[] = $child->getId();
+					}
+				} else {
+					/**
+					 * Before doing handover to transferShare, check if the share
+					 * id is present in the childShares. If so then just ignore
+					 * this share and continue.
```
Comment (PVince81) [inline apps/files/lib/Command/TransferOwnership.php]: I'm missing the "why" in this comment

Comment (sharidas) [reply apps/files/lib/Command/TransferOwnership.php]: Added the reason that processing child shares would result in processing the share again, while the share was processed along with the parent share. When the parent share was shared with the `destinationUser` then the child shares are processed. And hence we can ignore the processing of child shares further, if they are already processed.

**Comment by sharidas** [comment]
With the latest changes below are the tests done
- [x] The issue https://github.com/owncloud/core/issues/31150#issue-314806286 is solved
- [x] In the same test above, created 2 more users `user3` and `user4`. Now Tried to share the test folder to `user3` and `user4` and created public link for test folder for the new users. Tried to modify the parent column for the shares made for `user3` and `user4`, so that parents of initial share to `user2` be the same for `user3` and `user4`. Now ran the command. All the public link shares are accessible.
- [x] user1 creates a file which is shared with user3. Ran transfer-ownership command from user1 to user2. The shared file in user2 is still accessible.
- [x] The above test is conducted with sharing folder, and it works.
- [x] user3 is added to group1. user1 and user2 are not added to any group. user1 creates folder foo. And creates file foo/bar.txt. user1 shares foo to `group1` and `user3`. The transfer ownership command is executed from user1 to user2. The shares are available at user3. They are accessible from web ui.
- [x] Received shares are not transferred
- [x] user1 shares folder /foo/bar to user3. Delete /foo/bar from user1 to trashbin. Transfer ownership when ran from user1 to user2. It doesn't fail.
- [x] User1 creates foo/test.txt and shares `foo` to user2. Executes transfer ownership with path foo, the foo/test.txt is still available in `user2` under transferred folder.

**Review by PVince81 — APPROVED** [review]
:+1:

**Comment by PVince81** [comment]
@sharidas please backport

**Comment by sharidas** [comment]
Backport PR: https://github.com/owncloud/core/pull/31953
